presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -186,32 +186,23 @@ public synchronized void freeSpill(long bytes)
     public synchronized void setMemoryPool(MemoryPool pool)
     {
         requireNonNull(pool, "pool is null");
-        if (pool.getId().equals(memoryPool.getId())) {
+        if (memoryPool == pool) {
             // Don't unblock our tasks and thrash the pools, if this is a no-op
             return;
         }
         MemoryPool originalPool = memoryPool;
-        long originalReserved = reserved + revocableReserved;
         memoryPool = pool;
-        ListenableFuture<?> future = pool.reserve(queryId, originalReserved);
-        Futures.addCallback(future, new FutureCallback<Object>()
-        {
-            @Override
-            public void onSuccess(Object result)
-            {
-                originalPool.free(queryId, originalReserved);
-                // Unblock all the tasks, if they were waiting for memory, since we're in a new pool.
-                taskContexts.values().forEach(TaskContext::moreMemoryAvailable);
-            }
-
-            @Override
-            public void onFailure(Throwable t)
-            {
-                originalPool.free(queryId, originalReserved);
-                // Unblock all the tasks, if they were waiting for memory, since we're in a new pool.
-                taskContexts.values().forEach(TaskContext::moreMemoryAvailable);
-            }
-        });
+
+        ListenableFuture<?> memoryFuture = pool.reserve(queryId, reserved);
+        originalPool.free(queryId, reserved);
+
+        ListenableFuture<?> revocableMemoryFuture = pool.reserveRevocable(queryId, revocableReserved);
+        originalPool.freeRevocable(queryId, revocableReserved);
+
+        Futures.allAsList(memoryFuture, revocableMemoryFuture).addListener(() -> {
+            // Unblock all the tasks, if they were waiting for memory, since we're in a new pool.
+            taskContexts.values().forEach(TaskContext::moreMemoryAvailable);
```

Comment:
This listener and `TaskContext::moreMemoryAvailable` will be called when *both* `memoryFuture` and `revocableMemoryFuture` are completed. And `TaskContext::moreMemoryAvailable` eventually completes `OperatorContext.memoryFuture`, which is for the user memory. There is also a separate `revocableMemoryFuture` in `OperatorContext` for revocable memory. So why do we need the `revocableMemoryFuture` to complete for completing `OperatorContext.memoryFuture`, which is the future for user memory?

Reply:
You're right, fixing